vm/ubpf_vm.c

```diff
@@ -1188,31 +1198,53 @@ bounds_check(
 {
     if (!vm->bounds_check_enabled)
         return true;
-    if (mem && (addr >= mem && ((char*)addr + size) <= ((char*)mem + mem_len))) {
-        /* Context access */
-        return true;
-    } else if (addr >= stack && ((char*)addr + size) <= ((char*)stack + UBPF_STACK_SIZE)) {
-        /* Stack access */
-        return true;
-    } else if (
-        vm->bounds_check_function != NULL &&
-        vm->bounds_check_function(vm->bounds_check_user_data, (uintptr_t)addr, size)) {
-        /* Registered region */
-        return true;
-    } else {
+
+    uintptr_t access_start= (uintptr_t)addr;
+    uintptr_t access_end = access_start + size;
+    uintptr_t stack_start = (uintptr_t)stack;
+    uintptr_t stack_end = stack_start + UBPF_STACK_SIZE;
+    uintptr_t mem_start = (uintptr_t)mem;
+    uintptr_t mem_end = mem_start + mem_len;
+
+    if (access_start > access_end) {
         vm->error_printf(
             stderr,
-            "uBPF error: out of bounds memory %s at PC %u, addr %p, size %d\nmem %p/%zd stack %p/%d\n",
+            "uBPF error: invalid memory access %s at PC %u, addr %p, size %d\n",
             type,
             cur_pc,
             addr,
-            size,
-            mem,
-            mem_len,
-            stack,
-            UBPF_STACK_SIZE);
+            size);
         return false;
     }
+
+    // Check if the access is within the memory bounds.
+    if (access_start >= mem_start && access_end <= mem_end) {
```

Comment:
Just wanted to confirm that here (and below) that we are really interested in doing a `<=`.

Reply:
I think so? Both access_end and mem_end are pointing to the first byte after the valid region.

A valid access would be [mem_start, mem_end), so access_start == mem_start and access_end == mem_end is valid.

Reply:
I am still a little confused especially with the open interval at the end. I am going to write it out to help myself:

`mem_start` = $500$
`mem_len` = $75$
`mem_end` = `mem_start + mem_len` = $500 + 75$

The $75$ means that there are $75$ valid bytes and they are between $500$ and $574$ (inclusive). But, `mem_end` would be $575$. So, I would want `access_end` to be `<` `mem_end`? Of course, if `access_end` is calculated the same way as `mem_end`, then perhaps the `<=` is correct?

Reply:
In your example, assume a BPF program is accessing 4 bytes starting at address 571 (which would be valid as it is accessing bytes at index 571, 572, 573, and 574).

access_start = 571
access_end = 575

access_end <= mem_end is therefore true, but access_end < mem_end is false.

Assume a BPF program is accessing 4 bytes starting at address 572 (which would be invalid as it is accessing bytes at index 572, 573, 574, and 575).

access_start = 572
access_end = 576

access_end <= mem_end is therefore false as well as access_end < mem_end.

Reply:
Yes, we are on the same page! Because the `access_end` and `mem_end` are calculated in the same way, then `<=` is perfect. Sorry for being dense.